### PR TITLE
Add plugin hooks required for the encrypted links plugin

### PR DESCRIPTION
**Pull request by adisbladis**
This is being used by https://github.com/adisbladis/nixops-encrypted-links

### CODE REVIEW

File: nixops/backends/__init__.py [hunk 1]
```diff
@@ -50,14 +51,20 @@ def __init__(self, name: str, config: nixops.resources.ResourceEval):
 class MachineState(nixops.resources.ResourceState):
     """Base class for NixOps machine state objects."""
 
+    defn: Optional[MachineDefinition]
```
Comment (grahamc) [inline nixops/backends/__init__.py:54]: Should `defn` and its setting, be part of `ResourceState`'s default implementation?

File: nixops/plugins/enums.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from __future__ import annotations
```
Comment (adisbladis) [inline nixops/plugins/enums.py:1]: Note to self: Delete this file.

File: nixops/deployment.py [hunk 3]
```diff
@@ -139,7 +142,7 @@ def machines(self) -> Dict[str, nixops.backends.MachineState]:
         }
 
     @property
-    def active(
+    def active_machines(
         self,
     ) -> Dict[str, nixops.backends.MachineState]:  # FIXME: rename to "active_machines"
```
Comment (grahamc) [inline nixops/deployment.py:147]: Let's mark that FIXME as a FIXEDYOU, and move this rename and its fallout in to a separate commit near the beginning of the series?

File: nixops/deployment.py [hunk 4]
```diff
@@ -551,21 +554,16 @@ def get_arguments(self) -> Any:
...
         }
-        authorized_keys: Dict[str, List[str]] = {
-            m.name: [] for m in active_machines.values()
-        }
-        kernel_modules: Dict[str, Set[str]] = {
-            m.name: set() for m in active_machines.values()
-        }
-        trusted_interfaces: Dict[str, Set[str]] = {
-            m.name: set() for m in active_machines.values()
-        }
```
Comment (grahamc) [inline nixops/deployment.py:568]: Let's make these cleanup bits a separate commit, saying they're already unused -- that'll make the addition of plugin hooks easier to parse out.

File: nixops/deployment.py [hunk 5]
```diff
@@ -1699,5 +1679,12 @@ def _load_modules_from(dir: str) -> None:
         importlib.import_module("nixops." + dir + "." + module[:-3])
 
 
+class DeploymentPlugin(Protocol):
+    def physical_spec(
+        self, d: Deployment
+    ) -> Dict[str, List[Dict[Tuple[str, ...], Any]]]:
```
Comment (grahamc) [inline nixops/deployment.py:1685]: I think we should probably give this type a name, like:

```
MachineSpecs = Dict[str, List[Dict[Tuple[str, ...], Any]]]
```

or something?

File: nixops/plugins/enums.py (hunk 2 above)
Comment (grahamc) [inline nixops/plugins/enums.py:1]: you asked me to say this file should be deleted :)

File: nixops/util.py [hunk 6]
```diff
@@ -125,8 +126,21 @@ class ImmutableValidatedObject:
 
     _frozen: bool
 
-    def __init__(self, **kwargs):
-        anno: Dict = self.__annotations__
+    def __init__(self, *args: ImmutableValidatedObject, **kwargs):
```
Comment (grahamc) [inline nixops/util.py:129]: Let's get some tests in for this change, and maybe another commit?

File: nixops/backends/__init__.py (hunk 1 above)
Comment (adisbladis) [reply nixops/backends/__init__.py:65]: Yes, I think so.

File: nixops/deployment.py (hunk 5 above)
Comment (adisbladis) [reply nixops/deployment.py:1702]: Yes, this type is incredibly awkward, it's so awkward in fact that I don't expect plugin authors to directly type these structures.

Two things I have considered instead:
- Simple dict:
``` python
{
  "networking": {"hostName": "foobar"}
}
```

- Abstract away in a class:
``` python
spec = Spec()
spec.add_option("networking.hostName", "foobar")
```

**Comment by adisbladis** [comment]
This PR has been rolled up in #1383
